### PR TITLE
Pv2 297 calendar period refactoring

### PR DESCRIPTION
**Pull request by carlbm**
**NOTE THAT YOU NEED TO UPDATE YOUR DATABASE**

### CODE REVIEW

File: src/SFA.DAS.Payments.Model.Core/SfaPeriodBaseClass.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace SFA.DAS.Payments.Model.Core
+{
+    public abstract class SfaPeriodBaseClass
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.Model.Core/SfaPeriodBaseClass.cs]: Removed base class

File: src/SFA.DAS.Payments.Messages.UnitTests/CalendarPeriodTest.cs [hunk 2]
```diff
@@ -1,158 +1,158 @@
-﻿using System;
-using NUnit.Framework;
-using SFA.DAS.Payments.Model.Core;
+﻿//using System;
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.Messages.UnitTests/CalendarPeriodTest.cs]: Removed both file and project

File: src/SFA.DAS.Payments.Model.Core/DeliveryPeriod.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public DeliveryPeriod Clone()
+        {
+            return (DeliveryPeriod) MemberwiseClone();
+        }
+
+        public string Identifier { get; set; }
+
+        public static DeliveryPeriod CreateFromAcademicYearAndPeriod(string academicYear, byte period)
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.Model.Core/DeliveryPeriod.cs]: Added unit tests

File: src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/IProviderPaymentsRepository.cs [hunk 4]
```diff
@@ -9,14 +9,14 @@ namespace SFA.DAS.Payments.ProviderPayments.Application.Repositories
     public interface IProviderPaymentsRepository
     {
         Task SavePayment(PaymentModel paymentData, CancellationToken cancellationToken = default(CancellationToken));
-        Task<List<PaymentModel>> GetMonthEndPayments(short collectionYear, byte collectionPeriodMonth, long ukprn, CancellationToken cancellationToken = default(CancellationToken));
-        Task<List<long>> GetMonthEndUkprns(short collectionYear, 
-                                            byte collectionPeriodMonth, 
-                                            CancellationToken cancellationToken = default(CancellationToken));
+        Task<List<PaymentModel>> GetMonthEndPayments(string collectionPeriodName, long ukprn,
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/IProviderPaymentsRepository.cs]: Year and Month are not unique for a collection period

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs [hunk 5]
```diff
@@ -16,11 +16,12 @@ public MonthEndService(IProviderPaymentsRepository providerPaymentsRepository)
             this.providerPaymentsRepository = providerPaymentsRepository;
         }
 
-        public async Task<List<PaymentModel>> GetMonthEndPayments(short collectionYear, byte collectionPeriod, long ukprn, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<List<PaymentModel>> GetMonthEndPayments(string collectionPeriodName,
+            long ukprn,
+            CancellationToken cancellationToken = default(CancellationToken))
         {
-            var payments = await providerPaymentsRepository.GetMonthEndPayments(collectionYear, collectionPeriod, ukprn, cancellationToken);
+            var payments = await providerPaymentsRepository.GetMonthEndPayments(collectionPeriodName, ukprn, cancellationToken);
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs]: year and month are not unique

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/ApprenticeshipContractType2EarningEventProcessorTest.cs [hunk 6]
```diff
@@ -116,9 +117,19 @@ public async Task TestHandleNormalEvent()
...
-            var paymentHistoryEntities = new[] { new PaymentHistoryEntity { CollectionPeriod = "1819-R02", DeliveryPeriod = "1819-R02"} };
+            var paymentHistoryEntities = new[] { new PaymentHistoryEntity
+            {
+                CollectionPeriod = CollectionPeriod.CreateFromAcademicYearAndPeriod("1819", 2),
+                DeliveryPeriod = DeliveryPeriod.CreateFromAcademicYearAndPeriod("1819", 2),
+            } };
 
-            mocker.Mock<IPaymentKeyService>().Setup(s => s.GeneratePaymentKey("9", 1, period)).Returns("payment key").Verifiable();
+            mocker.Mock<IPaymentKeyService>().Setup(s => s.GeneratePaymentKey("9", 1, It.IsAny<DeliveryPeriod>()))
+                .Callback<string, int, DeliveryPeriod>((_, __, delPeriod) =>
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/ApprenticeshipContractType2EarningEventProcessorTest.cs]: Probably :) I didn't know about that overload. Will investigate

File: src/SFA.DAS.Payments.ProviderPayments.Application/Mapping/ProviderPaymentsProfile.cs [hunk 7]
```diff
@@ -12,15 +11,14 @@ public class ProviderPaymentsProfile : Profile
...
-
             CreateMap<FundingSourcePaymentEvent, PaymentModel>()
                 .Include<EmployerCoInvestedFundingSourcePaymentEvent, PaymentModel>()
                 .Include<SfaCoInvestedFundingSourcePaymentEvent, PaymentModel>()
                 .Include<SfaFullyFundedFundingSourcePaymentEvent, PaymentModel>()
                 .ForMember(dest => dest.ExternalId, opt => opt.ResolveUsing(src => Guid.NewGuid()))
                 .ForMember(dest => dest.Id, opt => opt.Ignore())
-                .ForMember(dest => dest.CollectionPeriod, opt => opt.MapFrom(source => new CalendarPeriod(source.CollectionPeriod.Year, source.CollectionPeriod.Month)))
-                .ForMember(dest => dest.DeliveryPeriod, opt => opt.MapFrom(source => new CalendarPeriod(source.DeliveryPeriod.Year, source.DeliveryPeriod.Month)))
+                .ForMember(dest => dest.CollectionPeriod, opt => opt.MapFrom(source => source.CollectionPeriod))
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Mapping/ProviderPaymentsProfile.cs]: Changed to use Clone()

File: src/SFA.DAS.Payments.Application/Data/Configurations/PaymentModelConfiguration.cs [hunk 8]
```diff
@@ -17,19 +17,16 @@ public void Configure(EntityTypeBuilder<PaymentModel> builder)
             builder.Property(x => x.Amount).HasColumnName(@"Amount");
             builder.OwnsOne(p => p.CollectionPeriod, cp =>
             {
-                cp.Property(x => x.Month).HasColumnName(@"CollectionPeriodMonth").IsRequired();
                 cp.Property(x => x.Name).HasColumnName(@"CollectionPeriodName").IsRequired();
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.Application/Data/Configurations/PaymentModelConfiguration.cs]: Changed to use Clone()

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/ApprenticeshipContractType2EarningEventProcessorTest.cs [hunk 9]
```diff
@@ -165,7 +177,13 @@ public async Task TestNoEventProducedWhenZeroToPay()
 
             var paymentHistoryEntities = new PaymentHistoryEntity[0];
 
-            mocker.Mock<IPaymentKeyService>().Setup(s => s.GeneratePaymentKey("9", 1, period)).Returns("payment key").Verifiable();
+            mocker.Mock<IPaymentKeyService>().Setup(s => s.GeneratePaymentKey("9", 1, It.IsAny<DeliveryPeriod>()))
+                .Callback<string, int, DeliveryPeriod>((_, __, delPeriod) =>
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/ApprenticeshipContractType2EarningEventProcessorTest.cs]: Changed

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/ApprenticeshipContractType2EarningEventProcessorTest.cs [hunk 10]
```diff
@@ -287,7 +314,14 @@ public async Task TestSfaContributionIsCalculatedForZeroEarningRefunds(decimal?
             if (employerContribution2.HasValue)
                 paymentHistoryEntities.Add(CreatePaymentEntity(employerContribution2.Value, earningEvent, FundingSourceType.CoInvestedEmployer));
 
-            mocker.Mock<IPaymentKeyService>().Setup(s => s.GeneratePaymentKey(earningEvent.LearningAim.Reference, (int)OnProgrammeEarningType.Balancing, deliveryPeriod)).Returns("payment key").Verifiable();
+            mocker.Mock<IPaymentKeyService>()
+                .Setup(s => s.GeneratePaymentKey(earningEvent.LearningAim.Reference, (int)OnProgrammeEarningType.Balancing, It.IsAny<DeliveryPeriod>()))
+                .Callback<string, int, DeliveryPeriod>((_, __, delPeriod) =>
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/ApprenticeshipContractType2EarningEventProcessorTest.cs]: Changed

File: src/SFA.DAS.Payments.RequiredPayments.Application/Infrastructure/Configuration/RequiredPaymentsProfile.cs [hunk 11]
```diff
@@ -11,8 +10,8 @@ public RequiredPaymentsProfile()
         {
             CreateMap<PaymentHistoryEntity, Payment>()
                 .ForMember(dest => dest.Id, opt => opt.Ignore())
-                .ForMember(dest => dest.DeliveryPeriod, opt => opt.ResolveUsing(src => new CalendarPeriod(src.DeliveryPeriod)))
-                .ForMember(dest => dest.CollectionPeriod, opt => opt.ResolveUsing(src => new CalendarPeriod(src.CollectionPeriod)));
+                .ForMember(dest => dest.DeliveryPeriod, opt => opt.ResolveUsing(src => src.DeliveryPeriod))
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Infrastructure/Configuration/RequiredPaymentsProfile.cs]: Updated

File: src/SFA.DAS.Payments.AcceptanceTests.Components.Core/SFA.DAS.Payments.AcceptanceTests.Components.Core.csproj [hunk 12]
```diff
@@ -90,6 +90,7 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
   </ItemGroup>
   <ItemGroup>
+    <None Include="app.config" />
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.AcceptanceTests.Components.Core/SFA.DAS.Payments.AcceptanceTests.Components.Core.csproj]: Please remove this not needed in this project

File: src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/ProviderPaymentsRepository.cs [hunk 13]
```diff
@@ -18,34 +18,32 @@ public ProviderPaymentsRepository(IPaymentsDataContext paymentsDataContext)
...
 
-        public async Task<List<PaymentModel>> GetMonthEndPayments(short collectionYear, byte collectionPeriodMonth, long ukprn, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<List<PaymentModel>> GetMonthEndPayments(string collectionPeriodName, long ukprn,
+            CancellationToken cancellationToken = default(CancellationToken))
         {
             var payments = await paymentsDataContext
                 .Payment.Where(p => p.Ukprn == ukprn &&
-                            p.CollectionPeriod.Year == collectionYear &&
-                            p.CollectionPeriod.Month == collectionPeriodMonth)
+                            p.CollectionPeriod.Name == collectionPeriodName)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/ProviderPaymentsRepository.cs]: Do we have an index on Collection name in the database project to optimise this query?

Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Repositories/ProviderPaymentsRepository.cs]: Yes

File: src/SFA.DAS.Payments.AcceptanceTests.Components.Core/SFA.DAS.Payments.AcceptanceTests.Components.Core.csproj (hunk 12 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.AcceptanceTests.Components.Core/SFA.DAS.Payments.AcceptanceTests.Components.Core.csproj]: Not mine! But I've removed it

**Pull request merged**